flathub/update-flathub.ts

```diff
@@ -101,43 +100,69 @@ async function main() {
   )
 
   const releaseNotesJson = JSON.parse(releaseNotesStdOut)
-  const releaseNotesComponents = releaseNotesJson.body.split('\n')
+  const releaseNotesComponents: string[] = releaseNotesJson.body.split('\n')
 
-  const hpXmlJson = convert.xml2js(hpXml, { compact: false }) as Element
-  const releaseNotesElements: Element[] = []
+  // XML Modification with fast-xml-parser
+  hpXml = fs.readFileSync(xmlFilePath).toString()
+
+  const parserOptions = {
+    ignoreAttributes: false,
+    preserveOrder: true,
+    format: true, // Enable formatting
+    indentBy: '  ' // Use two spaces for indentation
+  }
+
+  const parser = new XMLParser(parserOptions)
+
+  const hpXmlJson = parser.parse(hpXml)
+
+  const builder = new XMLBuilder(parserOptions)
+
+  const releaseNotesElements: Record<string, Array<Record<string, string>>>[] =
+    [] // An array to hold generated <li> elements
 
   for (const [i, releaseComponent_i] of releaseNotesComponents.entries()) {
-    //update metainfo hpXml
     if (i === 0) continue
     if (!releaseComponent_i.startsWith('*')) continue
     if (releaseComponent_i.includes('http')) continue
-    const releaseNoteElement: Element = {
-      type: 'element',
-      name: 'li',
-      elements: [
-        {
-          type: 'text',
-          text: releaseComponent_i.slice(1) //remove the asterisk
-        }
+
+    const li = releaseComponent_i
+      .replace(/\n/g, '')
+      .replace(/\r/g, '')
+      .replace(/\t/g, '')
+      .slice(1)
+      .trim()
+
+    // Creating the <li> element (compatible with fast-xml-parser)
+    const releaseNoteElement = {
+      li: [
+        { '#text': li } // Directly set the text within the <li> element
       ]
     }
+
     releaseNotesElements.push(releaseNoteElement)
   }
 
-  const componentsTag = hpXmlJson.elements?.[0]
-  const releasesTag = componentsTag?.elements?.find(
-    (val) => val.name === 'releases'
-  ) //.releases.release.description.ul =
-  const releaseListTag =
-    releasesTag?.elements?.[0].elements?.[0].elements?.find(
-      (val) => val.name === 'ul'
-    )
+  const componentsTag = hpXmlJson[1].component
+  // Locate the 'releases' element within the 'componentsTag'
+  const releasesTag = componentsTag.find((val) => val.releases !== undefined)
+
+  // Proceed to find the <ul> element as before
+  const releaseListTag = releasesTag?.releases[0].release[0].description[1]
+
   if (releaseListTag === undefined) {
     throw 'releaseListTag ul undefined'
   }
-  releaseListTag.elements = releaseNotesElements
-  hpXml = convert.js2xml(hpXmlJson)
-  fs.writeFileSync(xmlFilePath, hpXml)
+
+  releaseListTag.ul = [...releaseNotesElements] // Set the <ul> element to the generated <li> elements
+
+  console.log('new releaseListTag = ', JSON.stringify(releaseListTag, null, 2))
+
+  const updatedHpXml = builder.build(hpXmlJson)
+
+  console.log('updatedHpXml = ', updatedHpXml)
```

Comment:
just double checking if you meant to leave the console logs in

Reply:
Yes, good to debug in case we have future issues like the previous one.
Also causes no harm 😬